pkg/client/apiutil/apimachinery.go

```diff
@@ -93,5 +102,8 @@ func createRestConfig(gvk schema.GroupVersionKind, baseConfig *rest.Config) *res
 	if cfg.UserAgent == "" {
 		cfg.UserAgent = rest.DefaultKubernetesUserAgent()
 	}
+	if cfg.ContentType == "" && localScheme.Recognizes(gvk) {
```

Comment:
So you copy the client-go scheme so this is only used for built-in types? Are you sure protobuf works for all built-in types on all kube server versions? Does upstream have any docs on this?

Reply:
Do you know what happens if we request protobuf for a type that doesn't support it, is there maybe some way to negotiate?

Reply:
> So you copy the client-go scheme so this is only used for built-in types?

 Yes, this is only used for built-in resources. For custom resources, `CRDs` can not support Protocol Buffers but `Aggregated API` can. You can see it in [this table](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility).

> Are you sure protobuf works for all built-in types on all kube server versions? Does upstream have any docs on this?

All built-in resources are guaranteed to implement Protocal Buffers:

- This [design doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/protobuf.md#schema)  for contributors says protobuf schema for each API group and version in k8s.io/api will be generated from the objects in that API group and version.
- [hack/update-generated-protobuf.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/update-generated-protobuf.sh) in Kubernetes repo will find all packages contains `+k8s:protobuf-gen=package` and generate protobuf schema for them. And all API definition package must contain a doc.go which has these comments:

```golang
// +k8s:deepcopy-gen=package
// +k8s:protobuf-gen=packag
```

Reply:
missed this comment.  I *think* this should probably be a bit less magic, since this ignores aggregated api servers or any future improvements to CRDs.  This is probably fine in the short run, but in the long run I think we want to check discovery or something to make sure that this is actually true.

Reply:
Understand, but currently there is no way to check if a resource type supports protobuf or not. I have added a `TODO` to describe this.